### DISCUSSION TITLE
NIN Mudra Sequencing

### DISCUSSION THREAD

**Discussion by Evolutious** [PrincessRTFM/XIVComboPlugin#72] [General]
Hi @PrincessRTFM! Been a minute since we last spoke, I hope things have been well with you!

Question; Do you know of a way to script out the sequencing of the NIN's mudra to achieve a certain Ninjutsu action?

What I am trying to do is have Ninjutsu change into Jin, Chi or Ten in appropriate sequence based on what final action I want Ninjutsu to change into.

For example, If I press my key bind for Ninjustu with no modifier (SHIFT/CTRL/ALT) then I want it to do the sequence: 'Ten > Chi > OriginalHook(Ninjustu)' which will result in executing Raiton. My issue so far is I do not see any method available to know which mudra ability has been "pressed" like you would in a lastComboMove; for example.

Any ideas?

**Comment by VariableVixen** [comment]
That would also require sending execute-action payloads to the server, which _might_ be doable with `/action` commands, but it would need to be sequenced and timed, which would be an issue. Frankly, I think you're looking at the wrong issue with your idea: if pressing a button triggers a sequence of actions, you can easily have your own sequence position tracker, so that wouldn't be the hard part. The hard part would be sending the instructions with the right timing.

It's also worth mentioning that the existing framework for replacing actions can't help with _any_ of this, since it only hooks the game's own action ID replacer - the same function that traits use to upgrade actions automatically - rather than anything that would detect or intercept an action being _used_, so it can't even help to trigger the sequence.

All of that said, there's a... reasonable facsimile that you _could_ probably arrange within this plugin's framework, although I'm also not sure about tracking the last part of the sequence. ImGui definitely has a way to detect when modifier keys are pressed, so it should be entirely possible to detect those within a combo replacer for Ninjutsu (the action) such that holding various modifiers would turn it into the mudra you need. You'd still need to handle your own sequence tracking, and there's also the issue of changing the held modifiers mid-combo because the user would still have to be pressing the assigned button to trigger each mudra.

Finally, just because _XCVX's_ framework can't help with any of the automatic action sending or intercepting Ninjutsu to trigger it, that doesn't mean it's not possible. There are like, three different plugins out there that intercept action use to change the target, or (in at least one case) the action itself. Those should help you devise a way to handle the intercept part, although I don't think any official plugins will be sending actions. There may be an unofficial one somewhere, but I'm not aware of anything that does it outside of intercepting and changing the ID for an action that user already used.

I'm sorry I couldn't be more help, but I hope this was at least of some use!

**Reply by Evolutious** [reply]
Absolutely! I appreciate the feedback and advise. Mostly I was not sure if you or one of the other XIVCombo forkers had found a way to do this so I was not trying to reinvent the wheel. I will look into your suggestions and see what I can determine. Thanks! :D
